### PR TITLE
feat(node-resolve): dedupe by package name

### PR DESCRIPTION
**Pull request by LarsDenBakker**
## Rollup Plugin Name: `node-resolve`

This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers: fixes https://github.com/rollup/plugins/issues/82

### Description

This makes the `dedupe` option also dedupe by package name. 

For example if you have your `dedupe` option set as: `['my-package']` and in your code you have imports like this:

```js
import MyPackage from 'my-package';
import foo from 'my-package/foo.js';
```

Then you would expect both to be deduplicated. Right now imports are only matched literally, so you would need to list all possible imports in the dedupe array. This gets unmaintainable quickly, especially if you take into account imports with and without file extensions.

With change `node-resolve` first checks if the import is matched literally, and then checks if the import's package name is matched. [npm package name rules](https://docs.npmjs.com/files/package.json) ensure we can safely extract the first part of the import as name, or first two if the first part is a package scope.

This isn't a breaking change since the existing deduplication behavior is maintained, people can still list both package names and full import paths. Unless people are depending on the current behavior that deep imports are not deduplicated, which I think is unlikely.

### CODE REVIEW

File: packages/node-resolve/src/index.js [hunk 1]
```diff
@@ -136,6 +136,14 @@ function resolveImportSpecifiers(importSpecifierList, resolveOptions) {
   return p;
 }
 
+function getPackageName(id) {
+  const split = id.split('/');
+  if (split[0][0] === '@') {
+    return `${split[0]}/${split[1]}`;
```
Comment (wuweiweiwu) [inline packages/node-resolve/src/index.js]: should this just be `id`?

Comment (LarsDenBakker) [reply packages/node-resolve/src/index.js]: No, this will turn an import of `@my-scope/my-package/foo.js`  into `@my-scope/my-package`.

Comment (shellscape) [reply packages/node-resolve/src/index.js]: @LarsDenBakker that's excellent info. could you add a code comment to that effect there?

Comment (LarsDenBakker) [reply packages/node-resolve/src/index.js]: Yeah makes sense, I've added comments to the function.

**Comment by LarsDenBakker** [comment]
@shellscape Thanks, I've updated the README let me know if you think it needs more changes.

**Comment by shellscape** [comment]
@LarsDenBakker I'm going to give this a few more days to let others weigh in if they'd like to. Really appreciate your patience on this one. It's a significant change to be sure.

**Pull request merged**